compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -1045,6 +1046,56 @@ trait Applications extends Compatibility {
       tree
   }
 
+  /** Post process all arguments to by-name parameters by removing any not-null
+   *  info that was used when typing them. Concretely:
+   *  If an argument corresponds to a call-by-name parameter, drop all
+   *  embedded not-null assertions of the form `x.$asInstanceOf[x.type & T]`
+   *  where `x` is a reference to a mutable variable. If the argument still typechecks
+   *  with the removed assertions and is still compatible with the formal parameter,
+   *  keep it. Otherwise issue an error that the call-by-name argument was typed using
+   *  flow assumptions about mutable variables and suggest that it is enclosed
+   *  in a `byName(...)` call instead.
+   */
+  private def postProcessByNameArgs(fn: TermRef, app: Tree)(given ctx: Context): Tree =
+    fn.widen match
+      case mt: MethodType if mt.paramInfos.exists.isInstanceOf[ExprType] =>
+        app match
+          case Apply(fn, args) =>
+            val dropNotNull = new TreeMap with
+              override def transform(t: Tree)(given Context) = t match
+                case AssertNotNull(t0) if t0.symbol.is(Mutable) => transform(t0)
+                case t: ValDef if !t.symbol.is(Lazy) => super.transform(t)
```

Comment:
I don't fully follow the logic here, but is this case redundant, given that the default below also recurses on the child nodes (i.e. calls `super.transform(t)`)?

Reply:
It's the mirror image of usedOutOfOrder. Strict vals and vars are in-order all other definitions are out of order. The comment of the next line refers to this. I add a pointer to `usedOutOfOrder` now that it is merged.